content/blog/go-embedding-interfaces/index.md

```diff
@@ -0,0 +1,373 @@
+---
+{
+    title: 'Embedding Interfaces in Go',
+    published: '2024-03-04',
+    authors: ['rusher2004'],
+    tags: ['go', 'testing'],
+    description: ""
+}
+---
+
+I think about testing a lot. I've come to enjoy writing the tests for the things I build more than the things themselves.
+A large reason why is because of how easy it is to do with the Go standard library, especially when it comes to mocking dependencies with the use of interfaces.
+Here I'd like to show an example of how I make use of embedding interfaces to quickly create mock 
+
+## Embedding in Go
+
+Go doesn't do inheritance, but you can take a little shortcut to borrow from other types when you want to.
+First, let's look at an example of embedding a struct:
+
+```go
+type MyStruct struct {
+	message string
+	*log.Logger
+}
+
+m := MyStruct{
+    "Hello, Unicorns!",
+    log.New(os.Stdout, "MyStruct: ", log.Ldate|log.Ltime),
+}
+
+m.Println(m.message)
+```
+
+We get to use the `log.Logger` struct methods on our struct without the extra verbosity of doing `m.Logger.Println` all over the place. Pretty handy!
+
+You can also embed interfaces. Let's say that you want to make use of something that should satisfy both the [error](https://pkg.go.dev/builtin#error) and [fmt.Stringer](https://pkg.go.dev/fmt#Stringer) interfaces.
+You could make a definition like this:
+
+```go
+type StringError interface {
+	String() string // fmt.Stringer
+	Error() string  // error
+}
+```
+
+It works, but it doesn't quite communicate our intent to implement those existing and widely used interfaces.
+To improve that, we can embed the interfaces.
+
+```go
+type StringError interface {
+	error
+	fmt.Stringer
+}
+```
+
+That's better. Now, developers familiar with common Go types will see what you're intending to do.
+Bonus points: your IDE will show you what those embedded interfaces define.
+
+#### error
+
+![](./error_interface_def.png)
+
+#### fmt.Stringer
+
+![](./stringer_interface_def.png)
+
+## CHANGE THIS HEADING
```

Comment:
We probably wanna change this heading before going live too ;)

Reply:
Ha whoops, I could have sworn I had.